### ISSUE TITLE
Feature request: trim_trailing_whitespace = modified

### ISSUE THREAD

**Issue by Nowaker** [editorconfig/editorconfig#366]
Please introduce a new option for `trim_trailing_whitespace` = `modified`.

This would mean a whitespace would be trimmed if and only if a line is new or modified. Existing lines in a file would not get trimmed. Reasons:

- `git diff` clarity - easier to review my own code before committing
- overall diff view clarity - easier to do code reviews
- `git blame` working universally good in all situations
- it's a default setting in editors that offer this option

Related: https://github.com/editorconfig/editorconfig/issues/238

**Comment by hgraeber** [comment]
Is anywhere specified what @trim_trailing_whitespace = true@ means?

My editorconfig plugin for qtcreatot does what @trim_trailing_whitespace = modified@ shall do.

Regarding git: Maybe it is better to use .gitattributes to fix this for certain file types, similar to crlf handling.

**Comment by Nowaker** [comment]
> Is anywhere specified what `trim_trailing_whitespace = true` means?

https://github.com/editorconfig/editorconfig/issues/238 is a discussion about documenting what it means. It was never decided. Introduction of `trim_trailing_whitespace = modified` would sure clarify that.

> Regarding git: Maybe it is better to use .gitattributes to fix this for certain file types, similar to crlf handling.

There's a massive number of tools that won't care about `.gitattributes`. I think GitHub, GitLab, Gerrit, and anything web based.

**Comment by hax** [comment]
Unfortunately, add new value to an exist property will never be backward-compatibility as my knowledge.

**Comment by flit** [comment]
To avoid compatibility issues, instead of modifying the existing `trim_trailing_whitespace` property definition, a new property like `trim_trailing_whitespace_scope` could be added that accepts one or more values in the set: `file` (default), `modified`, `non-whitespace`.

So you could set `trim_trailing_whitespace_scope = modified, non-whitespace`, for instance.

(This is somewhat related to #238.)

**Comment by ppalaga** [comment]
I find it useful to have an option to trim a line only if it is new or modified.

Ad compatibility: A separate `trim_trailing_whitespace_scope` attribute would change the meaning of the existing `trim_trailing_whitespace` either. Having to do such changes for the sake of evolution is OK, if we introduce versioning of the EditorConfig spec, see https://github.com/editorconfig/editorconfig/issues/356

**Comment by craigbarnes** [comment]
If you don't want trailing whitespace in your codebase, why not remove it all in a single commit and be done with it? If the problem is in reviewing a several thousand line diff, it should be a no-op if you view it with `git diff --ignore-space-change`.

**Comment by Nowaker** [comment]
> why not just remove it all in a single commit and be done with it? 

- many tools using `git blame` showing incorrect committer on the line (e.g. GitLab, GitHub)
- because you'll cause unneeded conflicts between master and many longer-term branches

**Comment by craigbarnes** [comment]
Yeah, the `git blame` and merge conflicts issues are very valid points.

I guess rather than "tracking changes" and trimming just the changed lines on-save, an implementation could trim individual lines as and when they're edited instead. Some implementations may already do it that way, but the ones I've looked at all seem to trim the whole file in bulk when saving to disk.

> Many editors do it just right

You mean some editors use this proposed behaviour by default? Which ones?

**Comment by xmo-odoo** [comment]
> You mean some editors use this proposed behaviour by default? Which ones?

Jetbrain IDEs. I downloaded IDEA CE, opened it on a random folder, and its default setting for "Strip trailing spaces on save" is "Modified Lines".

**Comment by Nowaker** [comment]
Any update on this?

**Comment by ioquatix** [comment]
> So you could set trim_trailing_whitespace_scope = modified, non-whitespace, for instance.

This would be perfect. I specifically prefer to keep leading indentation in my code, but want to trim whitespace AFTER code.
